File: typegate/tests/docs/how-tos/prog_deploy/prog_deploy.ts
// skip:start
// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0
// skip:end

import { Policy, t, typegraph } from "@typegraph/sdk/index.ts";
import { DenoRuntime } from "@typegraph/sdk/runtimes/deno.ts";
import * as path from "node:path";
import { BasicAuth, tgDeploy } from "@typegraph/sdk/tg_deploy.ts";

// Your typegraph
export const tg = await typegraph("example", (g) => {
  const deno = new DenoRuntime();
  const pub = Policy.public();

  g.expose(
    {
      sayHello: deno.import(t.struct({ name: t.string() }), t.string(), {
        module: "scripts/say_hello.ts",
        name: "sayHello",
      }),
    },
    pub,
  );
});
// skip:start
const cwd = Deno.args[0];
const PORT = Deno.args[1];
// skip:end

// Configure your deployment
let baseUrl = "<TYPEGATE_URL>";
let auth = new BasicAuth("<USERNAME>", "<PASSWORD>");
// skip:start
baseUrl = `http://localhost:${PORT}`;
auth = new BasicAuth("admin", "password");
// skip:end

const config = {
  typegate: {
    url: baseUrl,
    auth: auth,
  },
  typegraphPath: path.join(cwd, "path-to-typegraph.ts"),
  prefix: "",
  secrets: {},
  migrationsDir: path.join("prisma-migrations", tg.name),
  defaultMigrationAction: {
    create: true,
    reset: true, // allow destructive migrations
  },
};

// Deploy to typegate
const deployResult = await tgDeploy(tg, config);
// skip:next-line
console.log(deployResult);

File: typegate/tests/docs/how-tos/prog_deploy/prog_deploy.py
import os
from os import path
import sys

from typegraph.gen.exports.core import MigrationAction
from typegraph.graph.shared_types import BasicAuth
from typegraph.graph.tg_deploy import (
    TypegraphDeployParams,
    tg_deploy,
    TypegateConnectionOptions,
)
from typegraph.runtimes.deno import DenoRuntime

from typegraph import Graph, Policy, t, typegraph

# Your typegraph


@typegraph()
def example(g: Graph):
    deno = DenoRuntime()
    pub = Policy.public()

    g.expose(
        pub,
        sayHello=deno.import_(
            t.struct({"name": t.string()}),
            t.string(),
            module="scripts/say_hello.ts",
            name="sayHello",
        ),
    )


# Configure your deployment
cwd = sys.argv[1]
PORT = sys.argv[2]


def deploy():
    cwd = os.getcwd()

    base_url = f"http://localhost:{PORT}"
    auth = BasicAuth("admin", "password")

    config: TypegraphDeployParams = (
        TypegraphDeployParams(
            typegate=TypegateConnectionOptions(url=base_url, auth=auth),
            typegraph_path=os.path.join(cwd, "path-to-typegraph"),
            prefix="<prefix>",
            secrets={},
            migrations_dir=path.join("prisma-migrations", example.name),
            migration_actions=None,
            default_migration_action=MigrationAction(
                apply=True,
                reset=True,  # allow destructive migrations
                create=True,
            ),
        ),
    )

    # Deploy to typegate
    result = tg_deploy(example, config)  # pass your typegraph function name
    return result


res = deploy()
